engine/src/main/java/io/zeebe/engine/processing/bpmn/container/MultiInstanceBodyProcessor.java

```diff
@@ -166,6 +167,66 @@ public void onEventOccurred(
     eventSubscriptionBehavior.triggerBoundaryEvent(element, context);
   }
 
+  @Override
+  public void onChildActivating(
+      final ExecutableMultiInstanceBody multiInstanceBody,
+      final BpmnElementContext flowScopeContext,
+      final BpmnElementContext childContext) {
+
+    final var bodyInstance = stateBehavior.getElementInstance(flowScopeContext);
+    final var loopCounter = bodyInstance.getMultiInstanceLoopCounter();
+
+    readInputCollectionVariable(multiInstanceBody, childContext)
+        .flatMap(
+            collection -> {
+              // the loop counter starts by 1
+              final var index = loopCounter - 1;
+              if (index < collection.size()) {
+                final var item = collection.get(index);
+                return Either.right(item);
+              } else {
+                final var incidentMessage =
+                    String.format(
+                        "Expected to read the %d multiInstanceBody of the input collection but it contains only %d elements. The input collection might be modified while iterating over it.",
+                        index, collection.size());
+                final var failure = new Failure(incidentMessage, ErrorType.EXTRACT_VALUE_ERROR);
+                return Either.left(failure);
```

Comment:
Currently, there are no tests to verify that the incident is created (because the code is not used yet). I created a follow-up issue to add the tests: #6755.